### DISCUSSION TITLE
what is the relation between shadow command argument conf. xml (the primary xml) and its hosts arguments sub xml?

### DISCUSSION THREAD

**Discussion by GitZinger** [shadow/shadow#775] [Q&A] [answered]
**State your question concisely**
did not mean to spam you guys. but i am quite confused about this. look through the WiKi, did not find the answer.
Take mini example,
in that folder, we could run 
`shadow shadow.conf.xml
`
but inside this xml, there are the topology(nodes and edges) and a bunch of hosts, and each host associates itself with an argument xml file, where there are another topology(nodes and edges).
the top xml entail the sub ones?
or the sub ones supersede the top one? **in term of the topology or the nodes and edges wise not the plugin or hosts settings**
thank you so much!
**Describe your current understanding of Shadow as it relates to your question**
all hosts are running in the same node actually. 
**Describe which parts you are unsure about**
since every host is running in the same node in the top xml file. then IPs stuff or networking route won't work or not useful because they are running in the same local node(machine)?

**Comment by robgjansen** [comment, marked as answer]
In the [basic example](https://github.com/shadow/shadow/blob/master/resource/examples/shadow.config.xml), we have:
  - a `<topology>` section. These nodes and edges represent the Internet, i.e., ISPs and paths between them.
  - a number of `<host>` elements. These represent "machines" that are running in the simulation. They get attached to the nodes in the topology, i.e., the get attached to the Internet via one of the `<nodes>` in the topology part.
  - each host can run some number `<process>` - these are applications that can run on a machine, like [Tor](https://www.torproject.org/) or [TGen](https://github.com/shadow/tgen), etc.

Now, I think your question is about the tgen arguments, for example:
```
<process plugin="tgen" starttime="1" arguments="tgen.server.graphml.xml"/>
```

The reason this refers to **another** graphml file with `<node>` and `<edge>` elements is that TGen also just happens to **also** use the graphml format to specify it's own *configuration* options. The node and edges in the TGen configuration file are completely independent of Shadow. The TGen graphml file use the terms "nodes" and "edges" because that is the standard graphml format.

[More info on TGen is here.](https://github.com/shadow/tgen/blob/master/doc/TGen-Overview.md)

Is that clear?

**Comment by GitZinger** [comment]
> In the [basic example](https://github.com/shadow/shadow/blob/master/resource/examples/shadow.config.xml), we have:
> 
> * a `<topology>` section. These nodes and edges represent the Internet, i.e., ISPs and paths between them.
> * a number of `<host>` elements. These represent "machines" that are running in the simulation. They get attached to the nodes in the topology, i.e., the get attached to the Internet via one of the `<nodes>` in the topology part.
> * each host can run some number `<process>` - these are applications that can run on a machine, like [Tor](https://www.torproject.org/) or [TGen](https://github.com/shadow/tgen), etc.
> 
> Now, I think your question is about the tgen arguments, for example:
> 
> ```
> <process plugin="tgen" starttime="1" arguments="tgen.server.graphml.xml"/>
> ```
> 
> The reason this refers to **another** graphml file with `<node>` and `<edge>` elements is that TGen also just happens to **also** use the graphml format to specify it's own _configuration_ options. The node and edges in the TGen configuration file are completely independent of Shadow. The TGen graphml file use the terms "nodes" and "edges" because that is the standard graphml format.
> 
> [More info on TGen is here.](https://github.com/shadow/tgen/blob/master/doc/TGen-Overview.md)
> 
> Is that clear?

In the minimal example, not only the TGen has its own configuration, but Tor also has its own one.
I thought everything runs on top of the Shadow such that all the hosts run in the Shadow  topology's nodes and communicate through its edges. 
1. if all hosts with the argument xml file run outside the Shadow in other network structure like additional nodes and edges, what is the point of the Shadow network? No host actually runs inside Shadow on Shadow nodes.
2. How do the hosts with different argument configuration xml file communicate with each other? since they are in totally different nodes, and there is no edge across different xml file?
thank you so much!

**Comment by robgjansen** [comment]
> if all hosts with the argument xml file run outside the Shadow in other network structure like additional nodes and edges, what is the point of the Shadow network? No host actually runs inside Shadow on Shadow nodes.

None of the processes are running outside of the Shadow simulation environment. The point of Shadow is to directly execute existing applications like Tor and TGen, but to "intercept" all functionality that would normally go through the Linux kernel and out to the Internet and instead send that through Shadow's simulated kernel and simulated Internet.

Still, if we run Tor entirely in Shadow's simulated network, but we directly execute the existing Tor software, then we need to be able to tell the software how it should behave. This is why each process in the shadow.config.xml file has an "arguments" section; when we start the corresponding process (in Shadow's simulated environment), we pass in those arguments to that virtual Tor process to instruct it how to behave (run in relay mode, client mode, etc.).

> How do the hosts with different argument configuration xml file communicate with each other? since they are in totally different nodes, and there is no edge across different xml file?
thank you so much!

Imagine you set up your own little network in your office, with some routers and desktop machines with Ubuntu installed. Your routers assign your desktop machines IP addresses and route the packets between the machines for you. Then on each machine you start a Tor process and a TGen process, passing the appropriate arguments to those applications to cause them to run as expected. They communicate with other Tor/TGen process on the other machines in your little network through your router. But you need to know which host machine to run each process, and if you have multiple routers, you need to know which router to plug each of your machines into.

This is basically what Shadow does for you, but it does it all in software.

**Comment by GitZinger** [comment]
thank you so much! Dr. Janson! I really appreciate it!
